include/pybind11/pybind11.h

```diff
@@ -1955,25 +1955,54 @@ inline std::pair<decltype(internals::registered_types_py)::iterator, bool> all_t
     return res;
 }
 
-template <typename Iterator, typename Sentinel, bool KeyIterator, return_value_policy Policy>
+/* There are a large number of apparently unused template arguments because
+ * each combination requires a separate py::class_ registration.
+ */
+template <typename Access, return_value_policy Policy, typename Iterator, typename Sentinel, typename ValueType, typename... Extra>
 struct iterator_state {
     Iterator it;
     Sentinel end;
     bool first_or_done;
 };
 
-PYBIND11_NAMESPACE_END(detail)
+// Note: these helpers take the iterator by non-const reference because some
+// iterators in the wild can't be dereferenced when const. C++ needs the extra parens in decltype
+// to enforce an lvalue. The & after Iterator is required for MSVC < 16.9. ResultType cannot be
+// reused for result_type due to bugs in ICC, NVCC, and PGI. See #3293.
+template <typename Iterator, typename ResultType = decltype((*std::declval<Iterator &>()))>
```

Comment:
How about 
```suggestion
template <typename Iterator, typename SFINAE = decltype((*std::declval<Iterator &>()))>
```
so that it's self-documenting that this argument is here for SFINAE purposes? Otherwise it looks like it could be removed.

Reply:
Done, thanks!